library/core/src/hint.rs

```diff
@@ -818,3 +818,115 @@ pub fn select_unpredictable<T>(condition: bool, true_val: T, false_val: T) -> T
+///
+/// A prefetch is a *hint*, and may be ignored on certain targets or by the hardware.
+///
+/// Passing a dangling or invalid pointer is permitted: the memory will not
+/// actually be dereferenced, and no faults are raised.
+///
+/// # Examples
+///
+/// ```
+/// use std::hint::{Locality, prefetch_read_data};
+/// use std::mem::size_of_val;
+///
+/// // Prefetch all of `slice` into the L1 cache.
+/// fn prefetch_slice<T>(slice: &[T]) {
+///     // On most systems the cache line size is 64 bytes.
+///     for offset in (0..size_of_val(slice)).step_by(64) {
+///         prefetch_read_data(slice.as_ptr().wrapping_add(offset), Locality::L1);
+///     }
+/// }
+/// ```
+#[inline(always)]
+#[unstable(feature = "hint_prefetch", issue = "146941")]
+pub const fn prefetch_read_data<T>(ptr: *const T, locality: Locality) {
+    match locality {
+        Locality::NonTemporal => {
+            intrinsics::prefetch_read_data::<T, { Locality::NonTemporal as i32 }>(ptr)
+        }
+        Locality::L3 => intrinsics::prefetch_read_data::<T, { Locality::L3 as i32 }>(ptr),
+        Locality::L2 => intrinsics::prefetch_read_data::<T, { Locality::L2 as i32 }>(ptr),
+        Locality::L1 => intrinsics::prefetch_read_data::<T, { Locality::L1 as i32 }>(ptr),
+    }
+}
+
+/// Prefetch the cache line containing `ptr` for a future write.
+///
+/// A strategically placed prefetch can reduce cache miss latency if the data is accessed
+/// soon after, but may also increase bandwidth usage or evict other cache lines.
+///
+/// A prefetch is a *hint*, and may be ignored on certain targets or by the hardware.
+///
+/// Passing a dangling or invalid pointer is permitted: the memory will not
+/// actually be dereferenced, and no faults are raised.
+#[unstable(feature = "hint_prefetch", issue = "146941")]
+pub const fn prefetch_write_data<T>(ptr: *mut T, locality: Locality) {
+    match locality {
+        Locality::NonTemporal => {
+            intrinsics::prefetch_write_data::<T, { Locality::NonTemporal as i32 }>(ptr)
+        }
+        Locality::L3 => intrinsics::prefetch_write_data::<T, { Locality::L3 as i32 }>(ptr),
+        Locality::L2 => intrinsics::prefetch_write_data::<T, { Locality::L2 as i32 }>(ptr),
+        Locality::L1 => intrinsics::prefetch_write_data::<T, { Locality::L1 as i32 }>(ptr),
+    }
+}
+
+/// Prefetch the cache line containing `ptr` for a future read.
+///
+/// A strategically placed prefetch can reduce cache miss latency if the instructions are
+/// accessed soon after, but may also increase bandwidth usage or evict other cache lines.
+///
+/// A prefetch is a *hint*, and may be ignored on certain targets or by the hardware.
+///
+/// Passing a dangling or invalid pointer is permitted: the memory will not
+/// actually be dereferenced, and no faults are raised.
+#[unstable(feature = "hint_prefetch", issue = "146941")]
+pub const fn prefetch_read_instruction<T>(ptr: *const T, locality: Locality) {
```

Comment:
shouldn't this be `ptr: unsafe fn()` or something since some platforms have different data and instruction pointer sizes?

Reply:
On some platforms a function pointer doesn't point directly to the instruction bytes, but rather to a function descriptor, which consists of a pointer to the first instruction and some value that needs to be loaded into a register. On these platforms using `unsafe fn()` would be incorrect. Itanium is an example, but I know there are more architectures that do this.

Reply:
ok, but that doesn't mean `*const T` is correct.

Reply:
ultimately all you need is an address, so `*const T` seemed the simplest way of achieving that.

Reply:
Hmm, well in that case the current implementation would not do anything, the LLVM intrinsic accepts a `ptr`

https://llvm.org/docs/LangRef.html#llvm-prefetch-intrinsic

These are also some fairly niche targets, for which I'm assuming prefetching (generally a performance measure) isn't very relevant. So maybe having to resort to inline assembly in those cases is acceptible?

Reply:
@programmerjake do you see that ACP moving forward? Maybe I should remove the instruction prefetching for now here and add it when there is progress?

Reply:
if you need just `Code`, you can probably get away with just adding that extern type for now under the tracking issue I just created https://github.com/rust-lang/rust/issues/148768 for that ACP and let whoever implements the rest of that ACP just use `Code`. You can add them all now and wait on that tracking issue for stabilization. If it takes too long, this feature can be partially stabilized and leave the code prefetch stabilization for later.